### PR TITLE
feat(primitives): revamp schematic primitives

### PR DESCRIPTION
**Pull request by rohanku**

### CODE REVIEW

File: codegen/src/block/mod.rs [hunk 1]
```diff
@@ -10,19 +10,16 @@ pub mod layout;
...
     generics: syn::Generics,
     io: syn::Type,
-    #[darling(multiple)]
-    #[allow(unused)]
-    schematic: Vec<darling::util::Ignored>,
-    #[darling(multiple)]
-    #[allow(unused)]
-    layout: Vec<darling::util::Ignored>,
-    #[darling(default)]
-    flatten: bool,
```
Comment (rohanku) [inline codegen/src/block/mod.rs:25]: Need to implement flatten.

File: tests/src/schematic.rs [hunk 2]
```diff
@@ -158,52 +163,52 @@ fn internal_signal_names_preserved() {
...
 fn nested_node_naming() {
     let ctx = Context::new(ExamplePdkA);
-    let handle = ctx.generate_schematic(BufferNxM::new(5, 5, 5));
+    let handle = ctx.generate_pdk_schematic(BufferNxM::new(5, 5, 5));
     let cell = handle.cell();
 
     assert_eq!(
-        cell.data().bubbled_inv1.data().pmos.terminals().g.path(),
-        cell.data().bubbled_pmos_g.path()
+        cell.nodes().bubbled_inv1.nodes().pmos.terminals().g.path(),
```
Comment (rohanku) [inline tests/src/schematic.rs:170]: `.nodes()` and `.terminals()` are too verbose.

File: tests/src/shared/buffer/schematic.rs [hunk 3]
```diff
@@ -42,29 +44,27 @@ impl Schematic<ExamplePdkA> for Inverter {
         cell.connect(io.vss, nmos.s);
         Ok(InverterData {
             pmos_g: pmos.terminals().g,
-            pmos,
+            pmos: pmos.data(),
```
Comment (rohanku) [inline tests/src/shared/buffer/schematic.rs:47]: Should not have to do `.data()`.

File: tests/src/shared/vdivider/flattened.rs [hunk 4]
```diff
@@ -1,12 +1,15 @@
...
-use substrate::pdk::Pdk;
-use substrate::schematic::{CellBuilder, ExportsSchematicData, Instance, Schematic, SchematicData};
+use substrate::io::SchematicType;
+use substrate::pdk::{Pdk, PdkSchematic, ToSchema};
+use substrate::schematic::schema::Schema;
+use substrate::schematic::{CellBuilder, ExportsNestedNodes, Instance, Schematic, SchematicData};
 
 #[derive(Clone, Copy, Hash, PartialEq, Eq, Serialize, Deserialize, Block)]
-#[substrate(io = "VdividerIo", flatten)]
+#[substrate(io = "VdividerIo", kind = "block::Cell")]
```
Comment (rohanku) [inline tests/src/shared/vdivider/flattened.rs:12]: Add back `flatten` here.

File: tools/spectre/src/lib.rs [hunk 5]
```diff
@@ -142,44 +159,59 @@ impl Options {
...
-        _ctx: &SimulationContext,
+        _ctx: &SimulationContext<Spectre>,
     ) {
         self.set_ic_inner(SimSignal::ScirVoltage(key.clone()), value);
     }
 }
 
-impl SetInitialCondition<&NodePath, Decimal> for Options {
-    fn set_initial_condition(&mut self, key: &NodePath, value: Decimal, ctx: &SimulationContext) {
+impl SetInitialCondition<&NodePath, Decimal, Spectre> for Options {
```
Comment (rohanku) [inline tools/spectre/src/lib.rs:179]: Find a unified API for these type of option modifiers.

File: libs/scir/src/drivers.rs [hunk 6]
```diff
@@ -205,10 +205,10 @@ impl LibraryBuilder {
                 .entered();
 
         // Cannot validate blackbox cells.
-        if cell.contents().is_opaque() {
+        if cell.contents().is_blackbox() {
```
Comment (rahulk29) [inline libs/scir/src/drivers.rs:208]: blackbox should be within-cell. replace current blackbox with external module.

File: tests/src/schematic.rs [hunk 7]
```diff
@@ -84,7 +88,7 @@ fn can_generate_flattened_vdivider_array_schematic() {
     let vdiv3 = crate::shared::vdivider::flattened::Vdivider::new(20, 20);
     let vdivs = vec![vdiv1, vdiv2, vdiv3];
     let vdivider = crate::shared::vdivider::flattened::VdividerArray { vdividers: vdivs };
-    let RawLib { scir, conv: _ } = ctx.export_scir(vdivider).unwrap();
+    let RawLib { scir, conv: _ } = ctx.export_scir::<Spice, _>(vdivider).unwrap();
```
Comment (rohanku) [inline tests/src/schematic.rs:91]: Is there a way to get rid of the `_`?

File: libs/scir/src/lib.rs [hunk 8]
```diff
@@ -913,7 +906,8 @@ impl LibraryBuilder {
             } else {
                 instance_names.push(inst.name().clone());
             }
-            id = inst.cell();
+            todo!()
```
Comment (rahulk29) [inline libs/scir/src/lib.rs:909]: sketchy

File: substrate/src/pdk/primitives.rs [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
+    pub ports: Vec<ArcStr>,
+    /// The parameters to pass to the instance.
+    pub params: IndexMap<ArcStr, Expr>,
+}
+
+impl Hash for RawInstance {
+    fn hash<H: Hasher>(&self, state: &mut H) {
+        self.cell.hash(state);
+        self.ports.hash(state);
+        self.params.iter().collect::<Vec<_>>().hash(state);
```
Comment (rahulk29) [inline substrate/src/pdk/primitives.rs:30]: sketchy

File: libs/spice/src/parser/conv.rs [hunk 10]
```diff
@@ -120,49 +110,52 @@ impl<'a> ScirConverter<'a> {
...
-                    let prim = scir::PrimitiveDeviceKind::Res2 {
-                        pos: node(&res.pos, &mut cell),
-                        neg: node(&res.neg, &mut cell),
+                    let id = self.lib.add_primitive(Primitive::Res2 {
                         value: str_as_numeric_lit(&res.value)?,
-                    };
-                    cell.add_primitive(scir::PrimitiveDevice::new(&**res.name, prim));
+                    });
+                    let mut sinst = scir::Instance::new(&**res.name, id);
+                    sinst.connect("1", node(&res.pos, &mut cell));
```
[the remark this answers is not in the document]
Comment (rohanku) [reply libs/spice/src/parser/conv.rs:117]: Convert SCIR instance based on primitive value, and convert each primitive value.

File: libs/scir/src/drivers.rs [hunk 11]
```diff
@@ -226,23 +220,10 @@ impl LibraryBuilder {
             }
         }
 
-        for (_, instance) in contents.instances.iter() {
+        for (_, instance) in cell.instances.iter() {
             analyze_instance(self, &mut net_states, instance);
         }
 
-        for (_, device) in contents.primitives.iter() {
```
Comment (rohanku) [inline libs/scir/src/drivers.rs:233]: add way to verify primitive connectivity

File: substrate/src/context.rs [hunk 12]
```diff
@@ -33,13 +36,26 @@ use crate::pdk::layers::LayerContext;
...
-use crate::simulation::{HasSimSchematic, SimController, SimulationContext, Simulator, Testbench};
+use crate::simulation::{SimController, SimulationContext, Simulator, Testbench};
+
+#[derive(Clone)]
+pub struct Context {
+    pub(crate) inner: Arc<RwLock<ContextInner>>,
+    simulators: Arc<HashMap<TypeId, Arc<dyn Any + Send + Sync>>>,
+    executor: Arc<dyn Executor>,
+    /// A cache for storing the results of expensive computations.
+    pub cache: Cache,
```
Comment (rohanku) [inline substrate/src/context.rs:57]: Is this cache tested?

File: codegen/src/block/mod.rs (hunk 1 above)
Comment (rohanku) [reply codegen/src/block/mod.rs:25]: also double check to remove all flatten invocations from macros.

File: codegen/src/block/schematic.rs [hunk 13]
```diff
@@ -219,125 +214,3 @@ impl ToTokens for DataInputReceiver {
...
-                            }
-                        ));
-                        Ok(())
-                    }
-                }
-            }
-        });
-
-        let expanded = quote! {
-            #has_schematic
```
Comment (rahulk29) [inline codegen/src/block/schematic.rs:334]: schematic

File: codegen/src/block/schematic.rs [hunk 14]
```diff
@@ -219,125 +214,3 @@ impl ToTokens for DataInputReceiver {
...
-
-            #(#has_schematic_impls)*
```
Comment (rahulk29) [inline codegen/src/block/schematic.rs:336]: schematic_impls

**Review by rahulk29 — APPROVED** [review]

**Pull request merged**